## Summary:
I have collected data from Vivino.com using their REST API. I have logged my network traffic using the Google Chrome's dev tools and obtained a JSON response with information for different wines. Then, I filtered and selected the required wine information that will be useful for the project. The selected variables are as follows: winery, wine_year, wine_rating, wine_price, wine_region, wine_country, grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin wine_description. Those features will be used in the EDA and during building Recommendation System. It should be noted that website response does not contain all features and for some wines all variables will be given, while for others it will be empty. 



In [1]:
# Importing Required Libraries
import requests
import csv
import pandas as pd
import numpy as np

In [11]:
'''
Function below has two entries (but it can be increased if you want to change API parameters at your convenience):

1) wine_type_id is a string type that represents type of wine available on Vivino: '1' is for Red Wine, '2' for White Wine, '3' for 
Sparkling Wine, '4' for Rose Wine, '7' for Dessert Wine and '24' for Fortified Wine.

2) price_increment is of an int type: It allows to change price increment of a API request after wine data has been collected
in for loop.

'''

def my_vivino_scraper(wine_type_id, price_increment):
    # Getting Request URL from the Website
    url = 'https://www.vivino.com/api/explore/explore?currency_code=USD&grape_filter=varietal&min_rating=1&order_by=price&order=asc&price_range_max=500&price_range_min=4.99'


    # Record the required Parameters, they might vary depending on your preferences
    params = {
                "country_code": "",
                "currency_code": "USD",
                "grape_filter": "varietal",
                "min_rating": "1", # Since Wine Ratings range from 1 to 5 I have set a minimum rating of 1 to get wines with wide range of ratings
                "order_by": "price", # Wines were sorted by Price
                "order": "asc", # Cheaper Wines appeared first
                "page": 1,   # Represents page: 1 page contains 25 wines 
                "price_range_max": "50000", # Maximum Wine Price Selected
                "price_range_min": 4.99, # Minimum Wine Price Selected
                "wine_type_ids[]": wine_type_id # Wines in Vivino has 6 different types 
    }

    # The User Agent was taken from Vivino Website response in the Google Chrome Dev Tools. 
    headers = {
        "user-agent": "Mozilla/5.0"
    }

    # Requests access to Vivino website using parameters described above
    response = requests.get(url, params=params, headers=headers)
    response.raise_for_status()

    # Select section that has useful wine informations from the JSON output
    records = response.json()["explore_vintage"]["records"]

    # Create List for parameters that I want to obtain
    wine_name, winery, wine_description, wine_price, wine_year, wine_rating, wine_region, wine_country, grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin =  [], [], [], [], [], [], [], [], [], [], [], [], []


    '''
    The chunk of code below contains three for loops: The inner for loop appends defined list-parameters with appropriate data obtained 
    from the JSON output. It continues until loop reaches length of current JSON output (25 entries per page).
    Since one page only shows 25 wine entries, next for loop increments page of the website allowing to request next 25 wines.
    Final for loop iterates the minimum price of a wine to ensure that we will access wines with wide range of prices.
    '''
    
    for k in range (250):

        for i in range (250):

            for j in range(len(records)):
                # Append The variables below with JSON output
                wine_name.append(records[j]['vintage']['wine']['name'])
                wine_price.append(records[j]['price']['amount'])
                wine_year.append(records[j]['vintage']['year'])
                wine_rating.append(records[j]['vintage']['statistics']['ratings_average'])
                winery.append(records[j]['vintage']['wine']['winery']['name'])

                # Variables below require try and except approach since not all wines have information available. Therefore, to avoid error of appending nothing try and except is used 
                try:

                    grape_information.append(records[j]['vintage']['wine']['style']['grapes'][0]['name'])

                except:

                    grape_information.append(np.nan)

                try:

                    wine_acidity.append(records[j]['vintage']['wine']['taste']['structure']['acidity'])

                except:

                    wine_acidity.append(np.nan)

                try:

                    wine_intensity.append(records[j]['vintage']['wine']['taste']['structure']['intensity'])

                except:

                    wine_intensity.append(np.nan)

                try:

                    wine_sweetness.append(records[j]['vintage']['wine']['taste']['structure']['sweetness'])

                except:

                    wine_sweetness.append(np.nan)

                try:

                    wine_tannin.append(records[j]['vintage']['wine']['taste']['structure']['tannin'])

                except:

                    wine_tannin.append(np.nan)

                try:

                    wine_region.append(records[j]['vintage'] ['wine']['region']['name'])

                except:

                    wine_region.append(np.nan)

                try:

                    wine_country.append(records[j]['vintage']['wine']['region']['country']['name'])

                except:

                    wine_country.append(np.nan)

                try:
                    wine_description.append(records[j]['vintage']['wine']['style']['description'])

                except:

                    wine_description.append(np.nan)

            params['page'] += 1 
            response = requests.get (url, params = params, headers = headers)
            records = response.json()['explore_vintage']['records']

        params['price_range_min'] += price_increment
    
    return (wine_name, winery, wine_description, wine_price, wine_year, wine_rating, wine_region, wine_country, grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin)


 

# Red Wine 

In [8]:
# Get the wine data we are interested in and call our function (Red Wine - '1', Price Increment Per loop - 10)
wine_name, winery, wine_description, wine_price, wine_year, wine_rating, wine_region, wine_country,  grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin = my_vivino_scraper('1', 10)

# Store the variables in the dictionary format
my_dictionary = ({
    'wine name': wine_name,
    'winery': winery,
    'wine year': wine_year,
    'wine rating': wine_rating,
    'wine price': wine_price,
    'wine region': wine_region,
    'wine country': wine_country,
    'grape information': grape_information,
    'wine acidity': wine_acidity,
    'wine intensity': wine_intensity,
    'wine sweetness': wine_sweetness,
    'wine tannin': wine_tannin, 
    'wine description': wine_description
})

# Convert dictionary to Pandas DataFrame and delete all duplicates 
data = pd.DataFrame(dict([(k, pd.Series(v)) for k, v in my_dictionary.items()]))
data.drop_duplicates(inplace= True)

# Write results to our dataframe
data.to_csv('Red_wine.csv')

wine name                  winery  wine year  \
0                        Côtes du Rhône              Montalcour       2014   
1                    Cabernet Sauvignon  Gallo Family Vineyards       2018   
2                                Merlot  Gallo Family Vineyards       2015   
3                     Cabernet - Merlot             Yellow Tail       2016   
4                         Bin 40 Merlot              Lindeman's       2016   
..                                  ...                     ...        ...   
195                            Pinotage                Beaumont       2017   
196    Purisima Mountain Vineyard Syrah                 Beckmen       2017   
197                    Ribera del Duero       Finca Villacreces       2016   
198  Les Fiefs de Lagrange Saint-Julien        Château Lagrange       2014   
199                                 Red          Shvo Vineyards       2014   

     wine rating  wine price              wine region   wine country  \
0            3.4        5.00           Côtes-du-Rhône         France   
1            3.6        5.59               California  United States   
2            3.5        5.59               California  United States   
3            3.4        5.95  South Eastern Australia      Australia   
4            3.4        5.95  South Eastern Australia      Australia   
..           ...         ...                      ...            ...   
195          4.0       36.99               Walker Bay   South Africa   
196          4.1       36.99           Ballard Canyon  United States   
197          3.9       36.99         Ribera del Duero          Spain   
198          3.7       36.99             Saint-Julien         France   
199          4.2       36.99            Upper Galilee         Israel   

      grape information  wine acidity  wine intensity  wine sweetness  \
0          Shiraz/Syrah      3.482611        3.792743        1.415443   
1    Cabernet Sauvignon      2.842703        4.697890        1.850506   
2                Merlot      1.620626        3.832823        2.327318   
3    Cabernet Sauvignon      3.898815        4.941685        1.548283   
4                Merlot      1.750228        3.885053        1.628029   
..                  ...           ...             ...             ...   
195            Pinotage      3.116149        3.827951        1.476855   
196        Shiraz/Syrah      3.056360        4.460651        1.619383   
197         Tempranillo      3.706280        3.961251        1.581326   
198  Cabernet Sauvignon      4.221845        3.868328        1.594276   
199                 NaN           NaN             NaN             NaN   

     wine tannin                                   wine description  
0       3.364203  The Southern Rhône is situated in a large, spr...  
1       3.247603  Known as the king of red wine grapes, Cabernet...  
2       1.985121  California Merlots tend to more fruit-forward,...  
3       3.418870  Australian Bordeaux blends are known for power...  
4       2.053112  Although many most closely associate Shiraz wi...  
..           ...                                                ...  
195     3.343851  An incredibly interesting wine, Pinotage is a ...  
196     3.620568  Californian Syrah certainly isn't a wine for t...  
197     3.569809  Rioja may be the most famous region in Spain, ...  
198     3.893925  Saint-Julien offers balanced, age-worthy blend...  
199          NaN                                                NaN  

[200 rows x 13 columns]

# White Wine 

In [13]:
# Get the wine data we are interested in and call our function (White wine - '2', Price Increment Per loop - 10)
wine_name, winery, wine_description, wine_price, wine_year, wine_rating, wine_region, wine_country,  grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin = my_vivino_scraper('2', 10)

# Store the variables in the dictionary format
my_dictionary = ({
    'wine name': wine_name,
    'winery': winery,
    'wine year': wine_year,
    'wine rating': wine_rating,
    'wine price': wine_price,
    'wine region': wine_region,
    'wine country': wine_country,
    'grape information': grape_information,
    'wine acidity': wine_acidity,
    'wine intensity': wine_intensity,
    'wine sweetness': wine_sweetness,
    'wine tannin': wine_tannin, 
    'wine description': wine_description

})

# Convert dictionary to Pandas DataFrame and delete all duplicates 
data = pd.DataFrame(dict([(k, pd.Series(v)) for k, v in my_dictionary.items()]))
data.drop_duplicates(inplace= True)
data
# Write results to our dataframe
data.to_csv('White_wine.csv')

wine name                  winery wine year  \
0                                   Moscato  Gallo Family Vineyards      2016   
1                           Sauvignon Blanc  Gallo Family Vineyards      2017   
2                       Bin 85 Pinot Grigio              Lindeman's      2019   
3                         Bin 65 Chardonnay              Lindeman's      2017   
4                    Astica Sauvignon Blanc                Trapiche      2015   
..                                      ...                     ...       ...   
620         Pouilly-Fuissé Le Clos Monopole          Château Fuissé      2016   
621  Camp Meeting Ridge Vineyard Chardonnay                 Flowers      2013   
622              Chardonnay (Hyde Vineyard)                     HDV      2012   
623                         Chardonnay Four             Liquid Farm      2013   
624                   Steiner Hund Riesling              Nikolaihof      2009   

     wine rating  wine price              wine region   wine country  \
0            3.9        5.59               California  United States   
1            3.3        5.59               California  United States   
2            3.5        5.95  South Eastern Australia      Australia   
3            3.2        5.95  South Eastern Australia      Australia   
4            3.2        5.99                  Mendoza      Argentina   
..           ...         ...                      ...            ...   
620          4.4       79.97           Pouilly-Fuissé         France   
621          4.4       79.97             Sonoma Coast  United States   
622          4.3       79.97             Los Carneros  United States   
623          4.3       79.99          Sta. Rita Hills  United States   
624          4.0       79.99                   Wachau        Austria   

    grape information  wine acidity  wine intensity  wine sweetness  \
0                 NaN           NaN             NaN             NaN   
1     Sauvignon Blanc           NaN             NaN             NaN   
2          Pinot Gris      2.736555        2.982939        1.724204   
3          Chardonnay      3.205055        3.460357        1.938636   
4     Sauvignon Blanc           NaN             NaN             NaN   
..                ...           ...             ...             ...   
620        Chardonnay      2.869677        3.153957        1.883385   
621        Chardonnay      3.341709        4.615300        2.762171   
622        Chardonnay      2.898492        4.633958        2.407105   
623        Chardonnay      3.464207        4.424609        2.454313   
624          Riesling      4.984336        3.435186        1.471232   

     wine tannin                                   wine description  
0            NaN                                                NaN  
1            NaN  California is known primarily for its Cabernet...  
2            NaN                                               None  
3            NaN  The main styles are unoaked Chardonnays in the...  
4            NaN                                               None  
..           ...                                                ...  
620          NaN                                               None  
621          NaN  With over 100,000 acres in production, Chardon...  
622          NaN                                               None  
623          NaN  With over 100,000 acres in production, Chardon...  
624          NaN                                               None  

[617 rows x 13 columns]

# Sparkling Wine

In [ ]:
# Get the wine data we are interested in and call our function (Sparkling wine - '3', Price Increment Per loop - 10)
wine_name, winery, wine_description, wine_price, wine_year, wine_rating, wine_region, wine_country,  grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin = my_vivino_scraper('3', 10)

# Store the variables in the dictionary format
my_dictionary = ({
    'wine name': wine_name,
    'winery': winery,
    'wine year': wine_year,
    'wine rating': wine_rating,
    'wine price': wine_price,
    'wine region': wine_region,
    'wine country': wine_country,
    'grape information': grape_information,
    'wine acidity': wine_acidity,
    'wine intensity': wine_intensity,
    'wine sweetness': wine_sweetness,
    'wine tannin': wine_tannin, 
    'wine description': wine_description
})

# Convert dictionary to Pandas DataFrame and delete all duplicates 
data = pd.DataFrame(dict([(k, pd.Series(v)) for k, v in my_dictionary.items()]))
data.drop_duplicates(inplace= True)

# Write results to our dataframe
data.to_csv('Sparkling_wine.csv')

# Rose Wine

In [ ]:
# Get the wine data we are interested in and call our function (Rose Wine - '4', Price Increment Per loop - 15)
wine_name, winery, wine_description, wine_price, wine_year, wine_rating, wine_region, wine_country,  grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin = my_vivino_scraper('4', 15)

# Store the variables in the dictionary format
my_dictionary = ({
    'wine name': wine_name,
    'winery': winery,
    'wine year': wine_year,
    'wine rating': wine_rating,
    'wine price': wine_price,
    'wine region': wine_region,
    'wine country': wine_country,
    'grape information': grape_information,
    'wine acidity': wine_acidity,
    'wine intensity': wine_intensity,
    'wine sweetness': wine_sweetness,
    'wine tannin': wine_tannin, 
    'wine description': wine_description

})

# Convert dictionary to Pandas DataFrame and delete all duplicates 
data = pd.DataFrame(dict([(k, pd.Series(v)) for k, v in my_dictionary.items()]))
data.drop_duplicates(inplace= True)
data
# Write results to our dataframe
data.to_csv('Rose_wine.csv')

# Dessert Wine

In [ ]:
# Get the wine data we are interested in and call our function (Dessert Wine - '7', Price Increment Per loop - 15)
wine_name, winery, wine_description, wine_price, wine_year, wine_rating, wine_region, wine_country,  grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin = my_vivino_scraper('7', 15)

# Store the variables in the dictionary format
my_dictionary = ({
    'wine name': wine_name,
    'winery': winery,
    'wine year': wine_year,
    'wine rating': wine_rating,
    'wine price': wine_price,
    'wine region': wine_region,
    'wine country': wine_country,
    'grape information': grape_information,
    'wine acidity': wine_acidity,
    'wine intensity': wine_intensity,
    'wine sweetness': wine_sweetness,
    'wine tannin': wine_tannin, 
    'wine description': wine_description
})

# Convert dictionary to Pandas DataFrame and delete all duplicates 
data = pd.DataFrame(dict([(k, pd.Series(v)) for k, v in my_dictionary.items()]))
data.drop_duplicates(inplace= True)
data
# Write results to our dataframe
data.to_csv('Dessert_wine.csv')

# Fortified Wine

In [1]:
# Get the wine data we are interested in and call our function (Fortified Wine - '24', Price Increment Per loop - 15)
wine_name, winery, wine_description, wine_price, wine_year, wine_rating, wine_region, wine_country,  grape_information, wine_acidity, wine_intensity, wine_sweetness, wine_tannin = my_vivino_scraper('24', 15)

# Store the variables in the dictionary format
my_dictionary = ({
    'wine name': wine_name,
    'winery': winery,
    'wine year': wine_year,
    'wine rating': wine_rating,
    'wine price': wine_price,
    'wine region': wine_region,
    'wine country': wine_country,
    'grape information': grape_information,
    'wine acidity': wine_acidity,
    'wine intensity': wine_intensity,
    'wine sweetness': wine_sweetness,
    'wine tannin': wine_tannin, 
    'wine description': wine_description

})

# Convert dictionary to Pandas DataFrame and delete all duplicates 
data = pd.DataFrame(dict([(k, pd.Series(v)) for k, v in my_dictionary.items()]))
data.drop_duplicates(inplace= True)
data
# Write results to our dataframe
data.to_csv('Fortified_wine.csv')

# Combining All Wine Data into one DataFrame 

In [3]:
red_wine_df = pd.read_csv('Red_wine.csv',index_col = [0])

red_wine_df.drop_duplicates(inplace = True)
red_wine_df['wine type'] = 'Red'
red_wine_df = red_wine_df[['wine name', 'winery', 'wine year', 'wine rating', 'wine price', 'wine type', 
                           'wine region', 'wine country', 'grape information', 'wine acidity', 'wine intensity',
                          'wine sweetness', 'wine tannin',  'wine description']]
print(red_wine_df.info())
red_wine_df


<class 'pandas.core.frame.DataFrame'>
Int64Index: 10305 entries, 0 to 22749
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   wine name          10305 non-null  object 
 1   winery             10305 non-null  object 
 2   wine year          10305 non-null  object 
 3   wine rating        10305 non-null  float64
 4   wine price         10305 non-null  float64
 5   wine type          10305 non-null  object 
 6   wine region        10296 non-null  object 
 7   wine country       10296 non-null  object 
 8   grape information  9911 non-null   object 
 9   wine acidity       5576 non-null   float64
 10  wine intensity     9801 non-null   float64
 11  wine sweetness     9801 non-null   float64
 12  wine tannin        9801 non-null   float64
 13  wine description   6864 non-null   object 
dtypes: float64(6), object(8)
memory usage: 1.2+ MB
None


wine name            winery  \
0                                     Cabernet Sauvignon       Carta Vieja   
1                                                 Merlot       Carta Vieja   
2                                     Cabernet Sauvignon      Three Wishes   
3                                     Cabernet Sauvignon        Crane Lake   
4                                             Pinot Noir        Crane Lake   
...                                                  ...               ...   
22583                                         Estate Red        Arkenstone   
22586                                   Cornas 'Reynard'  Thierry Allemand   
22598  Beckstoffer Missouri Hopper Vineyard Cabernet ...        The Debate   
22599                                           Insignia     Joseph Phelps   
22749                         Reserve Cabernet Sauvignon         PlumpJack   

      wine year  wine rating  wine price wine type        wine region  \
0          2019          3.4        4.99       Red  Loncomilla Valley   
1          2019          3.4        4.99       Red  Loncomilla Valley   
2          N.V.          3.1        4.99       Red         California   
3          2016          3.4        4.99       Red         California   
4          2016          3.4        4.99       Red         California   
...         ...          ...         ...       ...                ...   
22583      2016          4.6      169.99       Red    Howell Mountain   
22586      2012          4.5      375.00       Red             Cornas   
22598      2012          4.6      195.00       Red           Oakville   
22599      2007          4.7      279.99       Red        Napa Valley   
22749      2009          4.7      450.00       Red           Oakville   

        wine country   grape information  wine acidity  wine intensity  \
0              Chile  Cabernet Sauvignon      3.043747        3.781781   
1              Chile              Merlot      2.020424        3.482722   
2      United States  Cabernet Sauvignon      3.206160        4.545627   
3      United States  Cabernet Sauvignon      3.014199        4.738935   
4      United States          Pinot Noir      3.405433        2.832203   
...              ...                 ...           ...             ...   
22583  United States  Cabernet Sauvignon           NaN        4.848333   
22586         France        Shiraz/Syrah           NaN        4.542016   
22598  United States  Cabernet Sauvignon           NaN        4.589830   
22599  United States  Cabernet Sauvignon           NaN        4.777280   
22749  United States  Cabernet Sauvignon           NaN        4.846574   

       wine sweetness  wine tannin  \
0            1.772746     3.164342   
1            1.949938     2.366192   
2            1.962006     3.569953   
3            1.743180     3.540428   
4            1.500866     2.147871   
...               ...          ...   
22583        1.185455     4.116818   
22586        1.535436     3.488779   
22598        1.992003     3.406146   
22599        1.732723     3.498302   
22749        1.626460     3.372023   

                                        wine description  
0      Cabernet Sauvignon is the most widely grown gr...  
1      Merlot is a staple of the wine producing regio...  
2      Known as the king of red wine grapes, Cabernet...  
3      Known as the king of red wine grapes, Cabernet...  
4      Pinot Noir has the well deserved reputation of...  
...                                                  ...  
22583                                                NaN  
22586                                                NaN  
22598                                                NaN  
22599                                                NaN  
22749                                                NaN  

[10305 rows x 14 columns]

In [4]:
white_wine_df = pd.read_csv('White_wine.csv',index_col = [0] )

white_wine_df.drop_duplicates(inplace = True)

white_wine_df['wine type'] = 'White'
white_wine_df = white_wine_df[['wine name', 'winery', 'wine year', 'wine rating', 'wine price', 'wine type',  
                           'wine region', 'wine country', 'grape information', 'wine acidity', 'wine intensity',
                          'wine sweetness', 'wine tannin',  'wine description']]
# white_wine_df.info()
white_wine_df

wine name                        winery  \
0                          Bin 85 Pinot Grigio                    Lindeman's   
1                            Bin 65 Chardonnay                    Lindeman's   
2                              Sauvignon Blanc  Woodbridge by Robert Mondavi   
3                                       Grillo                          Tola   
4                   Coastal Estates Chardonnay        Beaulieu Vineyard (BV)   
...                                        ...                           ...   
7719  Puligny-Montrachet 1er Cru 'Champ Canet'                Etienne Sauzet   
7720          Clos Sainte Hune Riesling Alsace                      Trimbach   
7721  Blanc de Valandraud No. 1 Bordeaux Blanc            Château Valandraud   
7722                      Chassagne-Montrachet                   Bruno Colin   
7724          Clos Sainte Hune Riesling Alsace                      Trimbach   

     wine year  wine rating  wine price wine type  \
0         2019          3.5        5.95     White   
1         2017          3.2        5.95     White   
2         2017          3.3        5.99     White   
3         2018          3.6        5.99     White   
4         2017          3.5        6.95     White   
...        ...          ...         ...       ...   
7719      2013          4.3      175.00     White   
7720      1995          4.5      395.00     White   
7721      2014          4.0       86.50     White   
7722      2017          4.2       79.99     White   
7724      2013          4.5      195.43     White   

                                   wine region   wine country  \
0                      South Eastern Australia      Australia   
1                      South Eastern Australia      Australia   
2                                   California  United States   
3                              Terre Siciliane          Italy   
4                                   California  United States   
...                                        ...            ...   
7719  Puligny-Montrachet 1er Cru 'Champ Canet'         France   
7720                                    Alsace         France   
7721                                  Bordeaux         France   
7722                      Chassagne-Montrachet         France   
7724                                    Alsace         France   

     grape information  wine acidity  wine intensity  wine sweetness  \
0           Pinot Gris      2.737233        2.982922        1.724406   
1           Chardonnay      3.204798        3.460279        1.938847   
2      Sauvignon Blanc      4.063097        2.953770        1.419617   
3             Malvasia      2.979659        3.463870        1.172446   
4           Chardonnay      2.986839        4.776081        2.910311   
...                ...           ...             ...             ...   
7719        Chardonnay      3.285686        3.566332        1.772265   
7720          Riesling      4.441244        3.116488        2.384468   
7721   Sauvignon Blanc      2.943649        3.812263        1.867314   
7722        Chardonnay      3.299258        3.938477        1.587363   
7724          Riesling      4.441244        3.116488        2.384468   

      wine tannin                                   wine description  
0             NaN                                                NaN  
1             NaN  The main styles are unoaked Chardonnays in the...  
2             NaN  California is known primarily for its Cabernet...  
3             NaN  Italy is responsible for some of the finest wi...  
4             NaN  With over 100,000 acres in production, Chardon...  
...           ...                                                ...  
7719          NaN                                                NaN  
7720          NaN  Seldom oaked, Alsatian Riesling are typically ...  
7721          NaN  Bordeaux is the largest wine producing region ...  
7722          NaN                                                NaN  
7724        

In [5]:
sparkling_wine_df = pd.read_csv('Sparkling_wine.csv',index_col = [0])

sparkling_wine_df.drop_duplicates(inplace = True)

sparkling_wine_df['wine type'] = 'Sparkling'
sparkling_wine_df = sparkling_wine_df[['wine name', 'winery', 'wine year', 'wine rating', 'wine price', 'wine type',
                           'wine region', 'wine country', 'grape information', 'wine acidity', 'wine intensity',
                          'wine sweetness', 'wine tannin',  'wine description']]

print(sparkling_wine_df.info())

sparkling_wine_df



<class 'pandas.core.frame.DataFrame'>
Int64Index: 2230 entries, 0 to 1095
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   wine name          2230 non-null   object 
 1   winery             2230 non-null   object 
 2   wine year          2230 non-null   object 
 3   wine rating        2230 non-null   float64
 4   wine price         2230 non-null   float64
 5   wine type          2230 non-null   object 
 6   wine region        2228 non-null   object 
 7   wine country       2228 non-null   object 
 8   grape information  1749 non-null   object 
 9   wine acidity       1731 non-null   float64
 10  wine intensity     1731 non-null   float64
 11  wine sweetness     1 non-null      float64
 12  wine tannin        0 non-null      float64
 13  wine description   1571 non-null   object 
dtypes: float64(6), object(8)
memory usage: 261.3+ KB
None


wine name         winery  \
0            Governor's Cuvée Stanford Brut Champagne  Weibel Family   
1                                 Cava Gran Brut Rosé    Campo Viejo   
2                Verdi Raspberry Sparkletini Spumante          Bosca   
3                             Verdi Peach Sparkletini          Bosca   
4                              Cava Gran Brut Reserva    Campo Viejo   
...                                               ...            ...   
1091                                   Brut Champagne   Dom Pérignon   
1092                                   Brut Champagne   Dom Pérignon   
1093         Millésime Champagne Grand Cru 'Verzenay'   Hugues Godmé   
1094  Brut Vintage Champagne (Extra Cuvée de Réserve)      Pol Roger   
1095  Brut Vintage Champagne (Extra Cuvée de Réserve)      Pol Roger   

     wine year  wine rating  wine price  wine type  \
0         N.V.          3.0        5.99  Sparkling   
1         N.V.          3.6        7.95  Sparkling   
2         N.V.          3.6        7.95  Sparkling   
3         N.V.          3.7        7.95  Sparkling   
4         N.V.          3.6        7.95  Sparkling   
...        ...          ...         ...        ...   
1091      1999          4.6      199.99  Sparkling   
1092      1992          4.5      450.00  Sparkling   
1093      2009          4.1       74.99  Sparkling   
1094      2009          4.3       78.66  Sparkling   
1095      2006          4.3       96.99  Sparkling   

                         wine region   wine country grape information  \
0                         California  United States        Chardonnay   
1                               Cava          Spain          Xarel-lo   
2                                NaN            NaN               NaN   
3                               Asti          Italy               NaN   
4                               Cava          Spain          Xarel-lo   
...                              ...            ...               ...   
1091                       Champagne         France        Chardonnay   
1092                       Champagne         France        Chardonnay   
1093  Champagne Grand Cru 'Verzenay'         France        Chardonnay   
1094                       Champagne         France        Chardonnay   
1095                       Champagne         France        Chardonnay   

      wine acidity  wine intensity  wine sweetness  wine tannin  \
0         2.923922        3.703530             NaN          NaN   
1         3.640720        2.933004             NaN          NaN   
2              NaN             NaN             NaN          NaN   
3              NaN             NaN             NaN          NaN   
4         3.493387        2.953652             NaN          NaN   
...            ...             ...             ...          ...   
1091      4.441510        3.995037             NaN          NaN   
1092      4.441510        3.995037             NaN          NaN   
1093      4.487732        4.007870             NaN          NaN   
1094      4.550522        4.015999             NaN          NaN   
1095      4.550522        4.015999             NaN          NaN   

                                       wine description  
0     The elegance of sparkling wine suits Northern ...  
1                                                   NaN  
2                                                   NaN  
3                                                   NaN  
4                                                   NaN  
...                                                 ...  
1091  While there are many sparkling wine regions ar...  
1092  While there are many sparkling wine regions ar...  
1093  While there are many sparkling wine regions ar...  
1094  While there are many sparkling wine regions ar...  
1095  While there are many sparkling wine regions ar...  

[2230 rows x 14 columns]

In [6]:
rose_wine_df =  pd.read_csv('Rose_wine.csv',index_col = [0])

rose_wine_df.drop_duplicates(inplace = True)

rose_wine_df['wine type'] = 'Rose'
rose_wine_df = rose_wine_df[['wine name', 'winery', 'wine year', 'wine rating', 'wine price', 'wine type',  
                           'wine region', 'wine country', 'grape information', 'wine acidity', 'wine intensity',
                          'wine sweetness', 'wine tannin',  'wine description']]

rose_wine_df

wine name                 winery wine year  wine rating  \
0      Zinfandel White            Canyon Road      2018          3.4   
1                 Rosé         Ménage à Trois      2018          3.5   
2                 Rosé            Hess Select      2019          3.8   
3       Rosé of Malbec                  Crios      2019          3.7   
4                 Rosé  Chateau Ste. Michelle      2018          3.6   
...                ...                    ...       ...          ...   
1266  Private Bin Rosé            Villa Maria      2016          3.6   
1272  Inspiration Rosé       Château de Berne      2017          4.0   
1466     Bordeaux Rosé        Château Auguste      2018          4.1   
1470              Rosé                Massaya      2018          3.7   
1471     Bordeaux Rosé           Mouton Cadet      2016          3.4   

      wine price wine type        wine region   wine country  \
0          10.00      Rose         California  United States   
1          10.49      Rose         California  United States   
2          10.95      Rose         California  United States   
3          10.95      Rose      Lujan de Cuyo      Argentina   
4          10.95      Rose      Yakima Valley  United States   
...          ...       ...                ...            ...   
1266       11.98      Rose        Marlborough    New Zealand   
1272       19.99      Rose  Côtes de Provence         France   
1466       16.79      Rose           Bordeaux         France   
1470       23.00      Rose       Bekaa Valley        Lebanon   
1471       11.41      Rose           Bordeaux         France   

     grape information  wine acidity  wine intensity  wine sweetness  \
0                  NaN           NaN             NaN             NaN   
1                  NaN           NaN             NaN             NaN   
2                  NaN           NaN             NaN             NaN   
3                  NaN           NaN             NaN             NaN   
4                  NaN           NaN             NaN             NaN   
...                ...           ...             ...             ...   
1266               NaN           NaN             NaN             NaN   
1272      Shiraz/Syrah      3.929884        2.498264        1.412894   
1466               NaN           NaN             NaN             NaN   
1470               NaN           NaN             NaN             NaN   
1471               NaN           NaN             NaN             NaN   

      wine tannin                                   wine description  
0             NaN                                                NaN  
1             NaN                                                NaN  
2             NaN                                                NaN  
3             NaN                                                NaN  
4             NaN                                                NaN  
...           ...                                                ...  
1266          NaN                                                NaN  
1272          NaN  No summer afternoon is complete without a litt...  
1466          NaN                                                NaN  
1470          NaN                                                NaN  
1471          NaN                                                NaN  

[1534 rows x 14 columns]

In [7]:
dessert_wine_df = pd.read_csv('Dessert_wine.csv',index_col = [0])

dessert_wine_df.drop_duplicates(inplace = True)

dessert_wine_df['wine type'] = 'Dessert'

dessert_wine_df = dessert_wine_df[['wine name', 'winery', 'wine year', 'wine rating', 'wine price', 'wine type',  
                           'wine region', 'wine country', 'grape information', 'wine acidity', 'wine intensity',
                          'wine sweetness', 'wine tannin',  'wine description']]
print(dessert_wine_df.info())

dessert_wine_df

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1022 entries, 0 to 726
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   wine name          1022 non-null   object 
 1   winery             1022 non-null   object 
 2   wine year          1022 non-null   object 
 3   wine rating        1022 non-null   float64
 4   wine price         1022 non-null   float64
 5   wine type          1022 non-null   object 
 6   wine region        1022 non-null   object 
 7   wine country       1022 non-null   object 
 8   grape information  616 non-null    object 
 9   wine acidity       559 non-null    float64
 10  wine intensity     559 non-null    float64
 11  wine sweetness     558 non-null    float64
 12  wine tannin        10 non-null     float64
 13  wine description   164 non-null    object 
dtypes: float64(6), object(8)
memory usage: 119.8+ KB
None


wine name                      winery  \
0                          Fruitscato Watermelon                    Barefoot   
1                               Sweet Walter Red                  Bully Hill   
2                             Sweet Walter White                  Bully Hill   
3                  Harvest Select Sweet Riesling       Chateau Ste. Michelle   
4                  Harvest Select Sweet Riesling       Chateau Ste. Michelle   
..                                           ...                         ...   
489                                    Sauternes       Château Raymond-Lafon   
695                     Vouvray Le Mont Moelleux                Domaine Huet   
724  La Chapelle de Lafaurie-Peyraguey Sauternes  Château Lafaurie-Peyraguey   
725                    Alvaréga Malvasia di Bosa         G. Battista Columbu   
726                 Sauternes (Grand Cru Classé)      Château Doisy-Vedrines   

    wine year  wine rating  wine price wine type       wine region  \
0        N.V.          4.0        5.99   Dessert        California   
1        N.V.          3.7        8.99   Dessert          New York   
2        N.V.          4.0        8.99   Dessert          New York   
3        2017          4.0        9.95   Dessert   Columbia Valley   
4        2018          3.9        9.99   Dessert   Columbia Valley   
..        ...          ...         ...       ...               ...   
489      2002          4.1       42.89   Dessert         Sauternes   
695      2017          4.3       39.98   Dessert           Vouvray   
724      2015          4.1       44.99   Dessert         Sauternes   
725      2014          3.8       33.71   Dessert  Malvasia di Bosa   
726      2012          4.0       39.00   Dessert         Sauternes   

      wine country grape information  wine acidity  wine intensity  \
0    United States               NaN           NaN             NaN   
1    United States               NaN           NaN             NaN   
2    United States               NaN           NaN             NaN   
3    United States          Riesling      2.899289        2.511005   
4    United States          Riesling      2.899289        2.511005   
..             ...               ...           ...             ...   
489         France   Sauvignon Blanc      4.369525        4.699751   
695         France               NaN           NaN             NaN   
724         France   Sauvignon Blanc      4.300760        4.520865   
725          Italy               NaN           NaN             NaN   
726         France   Sauvignon Blanc      4.389975        4.770081   

     wine sweetness  wine tannin wine description  
0               NaN          NaN              NaN  
1               NaN          NaN              NaN  
2               NaN          NaN              NaN  
3          3.641250          NaN              NaN  
4          3.641250          NaN              NaN  
..              ...          ...              ...  
489        4.912559          NaN              NaN  
695             NaN          NaN              NaN  
724        4.919000          NaN              NaN  
725             NaN          NaN              NaN  
726        4.892039          NaN              NaN  

[1022 rows x 14 columns]

In [8]:
fortified_wine_df = pd.read_csv('Fortified_wine.csv',index_col = [0])

fortified_wine_df.drop_duplicates(inplace = True)

fortified_wine_df['wine type'] = 'Fortified'
fortified_wine_df = fortified_wine_df[['wine name', 'winery', 'wine year', 'wine rating', 'wine price', 'wine type',  
                           'wine region', 'wine country', 'grape information', 'wine acidity', 'wine intensity',
                          'wine sweetness', 'wine tannin',  'wine description']]

print(fortified_wine_df.info())

fortified_wine_df

<class 'pandas.core.frame.DataFrame'>
Int64Index: 811 entries, 0 to 813
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   wine name          811 non-null    object 
 1   winery             811 non-null    object 
 2   wine year          811 non-null    object 
 3   wine rating        811 non-null    float64
 4   wine price         811 non-null    float64
 5   wine type          811 non-null    object 
 6   wine region        809 non-null    object 
 7   wine country       809 non-null    object 
 8   grape information  766 non-null    object 
 9   wine acidity       87 non-null     float64
 10  wine intensity     87 non-null     float64
 11  wine sweetness     87 non-null     float64
 12  wine tannin        3 non-null      float64
 13  wine description   8 non-null      object 
dtypes: float64(6), object(8)
memory usage: 95.0+ KB
None


wine name                   winery  \
0            Colheita Port (Single Year Tawny)          Smith Woodhouse   
1    Colheita Malmsey Madeira (Single Harvest)                 Blandy's   
2              Quinta do Panascal Vintage Port                  Fonseca   
3                                 Vintage Port        Quinta do Vesuvio   
4                                 Vintage Port                    Kopke   
..                                         ...                      ...   
807                    10 Years Old White Port  Quinta de Santa Eufémia   
809                  Late Bottled Vintage Port         Quinta do Crasto   
810      Pedro Ximenez 1827 Jerez-Xeres-Sherry                  Osborne   
811                    Lagrima Fine White Port                    Krohn   
813                       Don PX Pedro Ximenez              Toro Albalá   

    wine year  wine rating  wine price  wine type                wine region  \
0        2000          4.2       55.99  Fortified                      Porto   
1        1999          4.2       57.20  Fortified            Malmsey Madeira   
2        2005          4.2       57.99  Fortified                      Porto   
3        2012          4.2       59.95  Fortified                      Porto   
4        2005          4.3       59.99  Fortified                      Porto   
..        ...          ...         ...        ...                        ...   
807      N.V.          4.1       29.99  Fortified                      Porto   
809      2015          4.2       27.48  Fortified                      Porto   
810      N.V.          3.9       17.59  Fortified  Pedro Ximénez Sherry (PX)   
811      N.V.          3.8       12.52  Fortified                      Porto   
813      2018          4.1       20.09  Fortified           Montilla-Moriles   

    wine country grape information  wine acidity  wine intensity  \
0       Portugal  Touriga Nacional           NaN             NaN   
1       Portugal          Verdelho      3.959141        4.638202   
2       Portugal  Touriga Nacional           NaN             NaN   
3       Portugal  Touriga Nacional           NaN             NaN   
4       Portugal  Touriga Nacional           NaN             NaN   
..           ...               ...           ...             ...   
807     Portugal           Gouveio           NaN             NaN   
809     Portugal  Touriga Nacional           NaN             NaN   
810        Spain     Pedro Ximenez      2.871021        4.823324   
811     Portugal           Gouveio           NaN             NaN   
813        Spain     Pedro Ximenez      3.003268        4.811542   

     wine sweetness  wine tannin wine description  
0               NaN          NaN              NaN  
1          4.667680          NaN              NaN  
2               NaN          NaN              NaN  
3               NaN          NaN              NaN  
4               NaN          NaN              NaN  
..              ...          ...              ...  
807             NaN          NaN              NaN  
809             NaN          NaN              NaN  
810        4.946997          NaN              NaN  
811             NaN          NaN              NaN  
813        4.903861          NaN              NaN  

[811 rows x 14 columns]

In [9]:
wine_df = pd.concat([red_wine_df, white_wine_df, sparkling_wine_df, rose_wine_df, dessert_wine_df, fortified_wine_df])
wine_df.drop_duplicates(inplace = True )

wine_df['wine price'].apply(type).value_counts()

wine_df = wine_df.reset_index(drop = True)
wine_df.to_csv('wine_df.csv')

In [10]:
wine_df

wine name                   winery  \
0                         Cabernet Sauvignon              Carta Vieja   
1                                     Merlot              Carta Vieja   
2                         Cabernet Sauvignon             Three Wishes   
3                         Cabernet Sauvignon               Crane Lake   
4                                 Pinot Noir               Crane Lake   
...                                      ...                      ...   
21976                10 Years Old White Port  Quinta de Santa Eufémia   
21977              Late Bottled Vintage Port         Quinta do Crasto   
21978  Pedro Ximenez 1827 Jerez-Xeres-Sherry                  Osborne   
21979                Lagrima Fine White Port                    Krohn   
21980                   Don PX Pedro Ximenez              Toro Albalá   

      wine year  wine rating  wine price  wine type  \
0          2019          3.4        4.99        Red   
1          2019          3.4        4.99        Red   
2          N.V.          3.1        4.99        Red   
3          2016          3.4        4.99        Red   
4          2016          3.4        4.99        Red   
...         ...          ...         ...        ...   
21976      N.V.          4.1       29.99  Fortified   
21977      2015          4.2       27.48  Fortified   
21978      N.V.          3.9       17.59  Fortified   
21979      N.V.          3.8       12.52  Fortified   
21980      2018          4.1       20.09  Fortified   

                     wine region   wine country   grape information  \
0              Loncomilla Valley          Chile  Cabernet Sauvignon   
1              Loncomilla Valley          Chile              Merlot   
2                     California  United States  Cabernet Sauvignon   
3                     California  United States  Cabernet Sauvignon   
4                     California  United States          Pinot Noir   
...                          ...            ...                 ...   
21976                      Porto       Portugal             Gouveio   
21977                      Porto       Portugal    Touriga Nacional   
21978  Pedro Ximénez Sherry (PX)          Spain       Pedro Ximenez   
21979                      Porto       Portugal             Gouveio   
21980           Montilla-Moriles          Spain       Pedro Ximenez   

       wine acidity  wine intensity  wine sweetness  wine tannin  \
0          3.043747        3.781781        1.772746     3.164342   
1          2.020424        3.482722        1.949938     2.366192   
2          3.206160        4.545627        1.962006     3.569953   
3          3.014199        4.738935        1.743180     3.540428   
4          3.405433        2.832203        1.500866     2.147871   
...             ...             ...             ...          ...   
21976           NaN             NaN             NaN          NaN   
21977           NaN             NaN             NaN          NaN   
21978      2.871021        4.823324        4.946997          NaN   
21979           NaN             NaN             NaN          NaN   
21980      3.003268        4.811542        4.903861          NaN   

                                        wine description  
0      Cabernet Sauvignon is the most widely grown gr...  
1      Merlot is a staple of the wine producing regio...  
2      Known as the king of red wine grapes, Cabernet...  
3      Known as the king of red wine grapes, Cabernet...  
4      Pinot Noir has the well deserved reputation of...  
...                                                  ...  
21976                                                NaN  
21977                                                NaN  
21978                                                NaN  
21979                                                NaN  
21980                                                NaN  

[21981 rows x 14 columns]

## End of Data Collection Notebook:
So we have 21981 unique wines collected from Vivino REST API. We can see that there are more than 10,000 Red Wines, around 6,000 White Wines and the Rest are spread between Sparkling, Rose, Dessert and Fortified Wines. In the next notebook you will see my attempt to make an Exploratory Data Analysis on collected wine data.